### PR TITLE
add deployment scale controller

### PR DESCRIPTION
**Pull request by captainroy-hy**
add deployment scale controller
add ut latter
Signed-off-by: roy wang <seiwy2010@gmail.com>

### CODE REVIEW

File: pkg/controller/common/rollout/workloads/deployment_scale_controller.go [hunk 1]
```diff
@@ -0,0 +1,303 @@
...
+	err := s.fetchDeployment(ctx)
+	if err != nil {
+		s.rolloutStatus.RolloutRetry(err.Error())
+		// nolint:nilerr
+		return false, nil
+	}
+	newPodTarget := calculateNewBatchTarget(s.rolloutSpec, int(s.rolloutStatus.RolloutOriginalSize),
+		int(s.rolloutStatus.RolloutTargetSize), int(s.rolloutStatus.CurrentBatch))
+	// get the number of ready pod from deployment
+	// TODO: should we use the replica number when we shrink?
```
Comment (wonderflow) [inline pkg/controller/common/rollout/workloads/deployment_scale_controller.go:195]: what does it mean?

File: pkg/controller/common/rollout/workloads/deployment_scale_controller.go [hunk 2]
```diff
@@ -0,0 +1,303 @@
...
+	if currentBatch.MaxUnavailable != nil {
+		unavail, _ = intstr.GetValueFromIntOrPercent(currentBatch.MaxUnavailable,
+			util.Abs(int(s.rolloutStatus.RolloutTargetSize-s.rolloutStatus.RolloutOriginalSize)), true)
+	}
+	klog.InfoS("checking the scaling progress", "current batch", s.rolloutStatus.CurrentBatch,
+		"new pod count target", newPodTarget, "new ready pod count", readyPodCount,
+		"max unavailable pod allowed", unavail)
+	s.rolloutStatus.UpgradedReadyReplicas = int32(readyPodCount)
+	targetReached := false
+	// nolint
```
Comment (wonderflow) [inline pkg/controller/common/rollout/workloads/deployment_scale_controller.go:208]: nolint for what?

**Review by wonderflow — COMMENTED** [review]
ping @ryanzhang-oss please review

File: pkg/controller/common/rollout/workloads/controller.go [hunk 3]
```diff
@@ -104,3 +105,39 @@ func (c *cloneSetController) fetchCloneSet(ctx context.Context) error {
...
+	err := c.client.Get(ctx, c.targetNamespacedName, &workload)
+	if err != nil {
+		if !apierrors.IsNotFound(err) {
+			c.recorder.Event(c.parentController, event.Warning("Failed to get the Deployment", err))
+		}
+		return err
+	}
+	c.deploy = &workload
+	return nil
+}
```
Comment (ryanzhang-oss) [inline pkg/controller/common/rollout/workloads/controller.go:143]: Why are those here? Looks like they are only used by the deploymentScaler

File: pkg/controller/common/rollout/workloads/deployment_scale_controller.go (hunk 1 above)
Comment (ryanzhang-oss) [reply pkg/controller/common/rollout/workloads/deployment_scale_controller.go:195]: This is copied from the clonesetController. We use the "readyReplicas" to check if the scaleup succeeded or not. it seems that we should use "replicas" to check if we scale down

File: pkg/controller/common/rollout/workloads/deployment_scale_controller.go (hunk 2 above)
Comment (ryanzhang-oss) [reply pkg/controller/common/rollout/workloads/deployment_scale_controller.go:208]: if/else

**Review by ryanzhang-oss — COMMENTED** [review]
LGTM. Please see if you can run the deployment examples with this check-in. 
The next step is to make this more generic.

File: pkg/controller/common/rollout/workloads/controller.go (hunk 3 above)
Comment (captainroy-hy) [reply pkg/controller/common/rollout/workloads/controller.go]: Fixed. I merged these helper functions into DeploymentScaleController.

File: pkg/controller/common/rollout/workloads/deployment_scale_controller.go (hunk 1 above)
Comment (wonderflow) [reply pkg/controller/common/rollout/workloads/deployment_scale_controller.go:197]: yes, check `replicas` in scale down case is better

**Review by wonderflow — APPROVED** [review]
lgtm @ryanzhang-oss  please merge after check

**Review by ryanzhang-oss — APPROVED** [review]
